src/katgpucbf/vgpu/recv.py

```diff
@@ -0,0 +1,286 @@
+            beam_ants = items[2]  # TODO should this be used to flag bad data?
+            payload_size = items[3]
+            pol = user_data[0].pol
+            if payload_size != heap_bytes or timestamp < 0 or frequency < 0 or beam_ants < 0:
+                # Probably a metadata heap - ignore it
+                batch_stats[user_data[0].stats_base + _Statistic.METADATA_HEAPS] += 1
+                return
+            if timestamp % heap_timestamp_step != 0:
+                # Invalid timestamp
+                batch_stats[user_data[0].stats_base + _Statistic.BAD_TIMESTAMP_HEAPS] += 1
+                return
+            if frequency % n_channels_per_substream != 0 or frequency >= n_channels:
+                # Invalid frequency
+                batch_stats[user_data[0].stats_base + _Statistic.BAD_FREQUENCY_HEAPS] += 1
+                return
+            # Heap index on the time axis, from timestamp 0
+            heap_time_abs = timestamp // heap_timestamp_step
+            data[0].chunk_id = heap_time_abs // n_batches_per_chunk
+            # Position of this heap on the time axis, from the start of the chunk
+            heap_time = heap_time_abs % n_batches_per_chunk
+            # Position of this heap on the frequency axis
+            heap_freq = frequency // n_channels_per_substream
+            data[0].heap_index = pol * n_heaps_per_pol + heap_time * n_pol_substreams + heap_freq
+            data[0].heap_offset = data[0].heap_index * heap_bytes
+
+        return chunk_place_impl
+
+
+def make_stream_group(
+    layout: Layout,
+    data_ringbuffer: spead2.recv.asyncio.ChunkRingbuffer,
+    free_ringbuffer: spead2.recv.ChunkRingbuffer,
+    recv_affinity: int,
+    pol_labels: Sequence[str],
+) -> spead2.recv.ChunkStreamRingGroup:
+    """Create a stream group for receiving dual-polarised beam data.
+
+    The readers are not added to the streams.
+
+    Parameters
+    ----------
+    layout
+        Heap size and chunking parameters.
+    data_ringbuffer
+        Output ringbuffer to which chunks will be sent.
+    free_ringbuffer
+        Ringbuffer for holding chunks for recycling once they've been used.
+    recv_affinity
+        CPU core affinity for the worker thread.
+        Use -1 to indicate no affinity.
+    pol_labels
+        Prometheus labels to apply to the polarisations (must have length 2).
+    """
+    # Reference counters to make the labels exist before the first scrape
+    assert len(pol_labels) == N_POLS
+    for pol in pol_labels:
+        counters.labels(pol)
+
+    user_data = np.zeros(N_POLS, dtype=user_data_type.dtype)
+    user_data["pol"] = np.arange(N_POLS)
+    group = base_recv.make_stream_group(
+        layout=layout,
+        spead_items=[TIMESTAMP_ID, FREQUENCY_ID, BEAM_ANTS_ID, spead2.HEAP_LENGTH_ID],
+        max_active_chunks=MAX_CHUNKS,
+        data_ringbuffer=data_ringbuffer,
+        free_ringbuffer=free_ringbuffer,
+        affinity=[recv_affinity] * N_POLS,
```

Comment:
I know the data rates are low in this Rx, but is this... "fine"? 
- Or rather, how did you arrive at this conclusion? i.e. same core for both recv streams